### PR TITLE
Adapt `PostProcessManager` and `OutputManager` to support complex values

### PR DESCRIPTION
**Pull request by tomcaruso**
This is achieved by separating the complex fields into their real and imaginary parts, which can then be visualized by VTK (which does not support complex fields).

### CODE REVIEW

**Comment by tomcaruso** [comment]
For now, the tests don't pass because their is a bug in pyccel that happen during the installation

**Comment by yguclu** [comment]
> Reminder: For now, the tests don't pass because their is a bug in pyccel that happen during the installation

Is this still true?

**Comment by tomcaruso** [comment]
> > Reminder: For now, the tests don't pass because their is a bug in pyccel that happen during the installation
> 
> Is this still true?

Nope, It was the bug we solved by upgrading pyccel. This branch could be merged right now.

File: psydac/core/kernels.py [hunk 1]
```diff
@@ -391,7 +397,7 @@ def eval_fields_3d_weighted(nc1: int, nc2: int, nc3: int, f_p1: int, f_p2: int,
                                                                 span_2 - f_p2:1 + span_2,
                                                                 span_3 - f_p3:1 + span_3]
 
-                arr_fields[:, :, :, :] = 0.0
+                arr_fields[:, :, :, :] *= 0.0
```
Comment (yguclu) [inline psydac/core/kernels.py:400]: Was the original code not working in the complex case? That would sound very strange.

**Review by yguclu — CHANGES_REQUESTED** [review]

File: psydac/core/kernels.py (hunk 1 above)
Comment (tomcaruso) [reply psydac/core/kernels.py:400]: I thought that it was not working but I test and it seems to work. What do you think is the most efficient way to do it ?

Comment (tomcaruso) [reply psydac/core/kernels.py:400]: Oh I just found the issue, in fact I can use `=0.0` for a vector but not for a scalar. Thus, I have made the change for to much variable but the issue is still relevant in the case of a complex scalar with `scalar=0.0`.

File: examples/notebooks/Helmotz_non_periodic.ipynb
[the remark this answers is not in the document]
Comment (tomcaruso) [reply examples/notebooks/Helmotz_non_periodic.ipynb:1]: [48778bc](https://github.com/pyccel/psydac/pull/312/commits/48778bcc4be1fae2bb97038de75e37e18518c9af)

File: examples/notebooks/Helmotz_non_periodic.ipynb [hunk 2]
```diff
@@ -0,0 +1,348 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Solving Poisson's equation on non-periodic topological domains\n",
```
[the remark this answers is not in the document]
Comment (tomcaruso) [reply examples/notebooks/Helmotz_non_periodic.ipynb]: [48778bc](https://github.com/pyccel/psydac/pull/312/commits/48778bcc4be1fae2bb97038de75e37e18518c9af)

File: examples/notebooks/Helmotz_non_periodic.ipynb [hunk 3]
```diff
@@ -0,0 +1,348 @@
...
+    "# from sympde import Domain\n",
+    "# connectivity = [((0,1,1),(1,1,-1))]\n",
+    "# patches = [Omega_1,Omega_2]\n",
+    "# Omega = Domain.join(patches, connectivity, 'domain')\n",
+    "\n",
+    "# Example of a complex multi-patch domain\n",
+    "# Omega = build_pretzel()\n",
+    "Omega=Square()\n",
+    "# Simple visualization of the topological domain\n",
+    "# plot_domain(Omega, draw=False, isolines=True)"
```
[the remark this answers is not in the document]
Comment (tomcaruso) [reply examples/notebooks/Helmotz_non_periodic.ipynb]: It's different domain I tried but our actual solution only work in a square. So this PR remove all those comments
[48778bc](https://github.com/pyccel/psydac/pull/312/commits/48778bcc4be1fae2bb97038de75e37e18518c9af)

File: examples/notebooks/Helmotz_non_periodic.ipynb [hunk 4]
```diff
@@ -0,0 +1,348 @@
...
+    "from sympde.expr.expr          import Norm                       \n",
+    "from sympde.expr               import find, EssentialBC\n",
+    "\n",
+    "from sympde.topology import ScalarFunctionSpace\n",
+    "from sympde.topology import elements_of\n",
+    "from sympde.topology import NormalVector\n",
+    "from sympde.topology import Union\n",
+    "from sympy import pi, cos, sin, symbols, conjugate, exp, sqrt\n",
+    "\n",
+    "# Define the abstract model to solve Helmotz's equation using the manufactured solution method\n",
```
[the remark this answers is not in the document]
Comment (tomcaruso) [reply examples/notebooks/Helmotz_non_periodic.ipynb]: [48778bc](https://github.com/pyccel/psydac/pull/312/commits/48778bcc4be1fae2bb97038de75e37e18518c9af)

File: examples/notebooks/Helmotz_non_periodic.ipynb [hunk 5]
```diff
@@ -0,0 +1,348 @@
...
+   "source": [
+    "# Save the results using OutputManager\n",
+    "from psydac.api.postprocessing import OutputManager\n",
+    "import os\n",
+    "\n",
+    "os.makedirs('results_helmotz', exist_ok=True)\n",
+    "\n",
+    "Om = OutputManager(\n",
+    "    f'results_helmotz/space_info_{Omega.name}',\n",
+    "    f'results_helmotz/field_info_{Omega.name}',\n",
```
[the remark this answers is not in the document]
Comment (tomcaruso) [reply examples/notebooks/Helmotz_non_periodic.ipynb]: [48778bc](https://github.com/pyccel/psydac/pull/312/commits/48778bcc4be1fae2bb97038de75e37e18518c9af)

File: examples/notebooks/Helmotz_non_periodic.ipynb [hunk 6]
```diff
@@ -0,0 +1,348 @@
...
+   "cell_type": "code",
+   "execution_count": 183,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Save the results using OutputManager\n",
+    "from psydac.api.postprocessing import OutputManager\n",
+    "import os\n",
+    "\n",
+    "os.makedirs('results_helmotz', exist_ok=True)\n",
```
[the remark this answers is not in the document]
Comment (tomcaruso) [reply examples/notebooks/Helmotz_non_periodic.ipynb]: [48778bc](https://github.com/pyccel/psydac/pull/312/commits/48778bcc4be1fae2bb97038de75e37e18518c9af)

File: examples/notebooks/paraview_images/Helmotz_Square_Imag_Computed.png
[the remark this answers is not in the document]
Comment (tomcaruso) [reply examples/notebooks/paraview_images/Helmotz_Square_Imag_Computed.png:1]: [1d81eca](https://github.com/pyccel/psydac/pull/312/commits/1d81eca61529347ab902a8205c60665f14df1b5b)

File: examples/notebooks/paraview_images/Helmotz_Square_Imag_error.png
[the remark this answers is not in the document]
Comment (tomcaruso) [reply examples/notebooks/paraview_images/Helmotz_Square_Imag_error.png:1]: [1d81eca](https://github.com/pyccel/psydac/pull/312/commits/1d81eca61529347ab902a8205c60665f14df1b5b)

File: examples/notebooks/paraview_images/Helmotz_Square_Real_Computed.png
[the remark this answers is not in the document]
Comment (tomcaruso) [reply examples/notebooks/paraview_images/Helmotz_Square_Real_Computed.png:1]: [1d81eca](https://github.com/pyccel/psydac/pull/312/commits/1d81eca61529347ab902a8205c60665f14df1b5b)

File: examples/notebooks/paraview_images/Helmotz_Square_Real_error.png
[the remark this answers is not in the document]
Comment (tomcaruso) [reply examples/notebooks/paraview_images/Helmotz_Square_Real_error.png:1]: [1d81eca](https://github.com/pyccel/psydac/pull/312/commits/1d81eca61529347ab902a8205c60665f14df1b5b)

File: examples/notebooks/paraview_images/Helmotz_Square_Real_exact.png
[the remark this answers is not in the document]
Comment (tomcaruso) [reply examples/notebooks/paraview_images/Helmotz_Square_Real_exact.png:1]: [1d81eca](https://github.com/pyccel/psydac/pull/312/commits/1d81eca61529347ab902a8205c60665f14df1b5b)

File: psydac/api/postprocessing.py [hunk 7]
```diff
@@ -860,6 +860,8 @@ def _reconstruct_spaces(self):
                 for sc_sp in components:
                     already_used_names.append(sc_sp['name'])
                     basis += sc_sp['basis']
+                    # TODO change codomain type when implemanted in symPDE
```
[the remark this answers is not in the document]
Comment (tomcaruso) [reply psydac/api/postprocessing.py]: [b1de26a](https://github.com/pyccel/psydac/pull/312/commits/b1de26a80430f72c5ed1c582871ba71e5eba9c2b)

File: psydac/api/postprocessing.py [hunk 8]
```diff
@@ -881,11 +883,13 @@ def _reconstruct_spaces(self):
                         for j in range(new_degree[i] - degree[0][i]):
                             knots[i] = elevate_knots(knots[i], degree[0][i], periodic=periodic[i])
 
+                # TODO change codomain type when implemanted in symPDE
```
[the remark this answers is not in the document]
Comment (tomcaruso) [reply psydac/api/postprocessing.py]: [b1de26a](https://github.com/pyccel/psydac/pull/312/commits/b1de26a80430f72c5ed1c582871ba71e5eba9c2b)

File: psydac/api/postprocessing.py [hunk 9]
```diff
@@ -917,17 +921,21 @@ def _reconstruct_spaces(self):
 
                     knots = [np.asarray(sc_sp['knots'][i]) for i in range(sc_sp['ldim'])]
                     periodic = sc_sp['periodic']
+                    # TODO change codomain type when implemanted in symPDE
```
[the remark this answers is not in the document]
Comment (tomcaruso) [reply psydac/api/postprocessing.py]: [b1de26a](https://github.com/pyccel/psydac/pull/312/commits/b1de26a80430f72c5ed1c582871ba71e5eba9c2b)

File: psydac/api/postprocessing.py [hunk 10]
```diff
@@ -917,17 +921,21 @@ def _reconstruct_spaces(self):
...
+                    codomain_type = 'complex' if sc_sp['dtype'] == "<class 'complex'>" else 'real'
 
                     for i in range(sc_sp['ldim']):
                         if new_degree[i] != degree[i]:
                             for j in range(new_degree[i] - degree[i]):
                                 knots[i] = elevate_knots(knots[i], degree[i], periodic=periodic[i])
 
+                    # TODO change codomain type when implemanted in symPDE
```
[the remark this answers is not in the document]
Comment (tomcaruso) [reply psydac/api/postprocessing.py]: [b1de26a](https://github.com/pyccel/psydac/pull/312/commits/b1de26a80430f72c5ed1c582871ba71e5eba9c2b)

File: examples/maxwell_2d_multi_patch.py [hunk 11]
```diff
@@ -114,51 +115,83 @@ def run_maxwell_2d(uex, f, alpha, domain, ncells, degree, k=None, kappa=None, co
...
+    #
+    # my_small_plot(
+    #     title=r'approximation of solution $u$, $y$ component',
+    #     vals=[E_y_vals, Eh_y_vals, E_y_err],
+    #     titles=[r'$u^{ex}_y(x,y)$', r'$u^h_y(x,y)$', r'$|(u^{ex}-u^h)_y(x,y)|$'],
+    #     xx=xx,
+    #     yy=yy,
+    #     gridlines_x1=None,
+    #     gridlines_x2=None,
+    # )
```
[the remark this answers is not in the document]
Comment (tomcaruso) [reply examples/maxwell_2d_multi_patch.py]: Nope it was just a mistake I made. I commit some test for define a bug. I remove all change on this file in this commit
[ace7249](https://github.com/pyccel/psydac/pull/312/commits/ace72493b9481922c8aeb00a7829ee503101bc3e)

File: psydac/fem/tensor.py [hunk 12]
```diff
@@ -500,9 +500,9 @@ def eval_fields_regular_tensor_grid(self, grid, *fields, weights=None, overlap=0
         degree, global_basis, global_spans, local_shape = self.preprocess_regular_tensor_grid(grid, der=0, overlap=overlap)
         ncells = [local_shape[i][0] for i in range(self.ldim)]
         n_eval_points = [local_shape[i][1] for i in range(self.ldim)]
-        out_fields = np.zeros((*(tuple(ncells[i] * n_eval_points[i] for i in range(self.ldim))), len(fields)))
+        out_fields = np.zeros((*(tuple(ncells[i] * n_eval_points[i] for i in range(self.ldim))), len(fields)), dtype=fields[0].coeffs.dtype)
 
-        glob_arr_coeffs = np.zeros(shape=(*fields[0].coeffs._data.shape, len(fields)))
+        glob_arr_coeffs = np.zeros(shape=(*fields[0].coeffs._data.shape, len(fields)), dtype=fields[0].coeffs.dtype)
```
[the remark this answers is not in the document]
Comment (tomcaruso) [reply psydac/fem/tensor.py]: You are right I forgot the that TensorFemSpace have the dtype
[ace7249](https://github.com/pyccel/psydac/pull/312/commits/ace72493b9481922c8aeb00a7829ee503101bc3e)

File: psydac/core/kernels.py (hunk 1 above)
Comment (tomcaruso) [reply psydac/core/kernels.py]: In this commit, I removed all the `*=` when the lhs is an array
[a63c4f1](https://github.com/pyccel/psydac/pull/312/commits/a63c4f1336baa69f1ecd7ea0477a70a69b7c55fa)

File: pyproject.toml [hunk 13]
```diff
@@ -33,7 +33,7 @@ dependencies = [
 
     # Our packages from PyPi
     'sympde==0.17.0',
-    'pyccel>=1.7.3',
+    "pyccel@git+https://github.com/pyccel/pyccel.git",
```
[the remark this answers is not in the document]
Comment (yguclu) [reply pyproject.toml:36]: You can now use Pyccel 1.8.1:
```python
'pyccel >= 1.8.1'
```

[the remark this answers is not in the document]
Comment (yguclu) [reply pyproject.toml:36]: :wave: @tomcaruso

File: examples/notebooks/Helmholtz_non_periodic.ipynb [hunk 14]
```diff
@@ -0,0 +1,326 @@
...
+    ")\n",
+    "\n",
+    "Pm.close()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Example: Results on the pretzel domain\n",
```
Comment (yguclu) [inline examples/notebooks/Helmholtz_non_periodic.ipynb:261]: This is a square pretzel :rofl:

**Review by yguclu — CHANGES_REQUESTED** [review]

File: examples/notebooks/Helmholtz_non_periodic.ipynb (hunk 14 above)
Comment (tomcaruso) [reply examples/notebooks/Helmholtz_non_periodic.ipynb:261]: The famous one ^^'

Comment (yguclu) [reply examples/notebooks/Helmholtz_non_periodic.ipynb:261]: BTW this notebook has two empty cells at the bottom of the page. Please remove them :wink:

**Review by yguclu — APPROVED** [review]
Great job!

**Pull request merged**